### DISCUSSION TITLE
No server running on socket, but why?

### DISCUSSION THREAD

**Discussion by Seltyk** [Infinidoge/nix-minecraft#104] [Q&A] [answered]
Whenever I try to start the server, whether thhat's an autostart on boot or a manual `systemctl start` command, the service immediately deactivates. In journalctl, the only hint of what's wrong appears to be this line:

```
Oct 12 20:33:36 irukandji minecraft-server-vanilla-stop[1016]: no server running on /run/minecraft/vanilla.sock
```

I've verified that `/run/minecraft` and the `*.sock` files within it are all owned by `minecraft:minecraft`. What follows is (what I hope to be) the only relevant part of the NixOS config.

```nix
    servers.vanilla = {
      enable = true;
      package = pkgs.vanillaServers.vanilla;
      autoStart = true;
      jvmOpts = builtins.concatStringsSep
        " "
        [
          "-XX:+CMSClassUnloadingEnabled"
          "-XX:+CMSIncrementalPacing"
          "-XX:+UseG1GC"
          "-XX:MaxHeapFreeRatio=10"
          "-XX:MinHeapFreeRatio=5"
          "-XX:ParallelGCThreads=2"
          "-Xms1024M"
          "-Xmx6144M"
        ];
      openFirewall = true;
      # prevent systemd from complaining about start-limit-hit
      restart = "no";
      serverProperties = {
        difficulty = "normal";
        motd = "Seltyk's vanilla server";
        server-port = 22113;
      };
    };
```

**Comment by Seltyk** [comment, marked as answer]
Update: the two CMS JVM options do not work. Commenting them out (and, at least temporarily, using systemd-socket instead of tmux) could at least get the server to start and operate.

**Reply by olivia-fl** [reply]
So I think the real bug/issue here in `nix-minecraft` is that if you're using tmux then error messages from crashes that occur early in startup are not visible anywhere.

**Reply by Seltyk** [reply]
Having Java errors buried in tmux is certainly an issue, so thank you for highlighting it. I only wrote the update comment in case someone else trying to use those flags came along wondering why Java wasn't starting for them.

**Reply by Infinidoge** [reply]
Yep, Benjamin is correct that the main issue with nix-minecraft's use of tmux is that it obfuscates a lot of this stuff. Some day I want to make a replacement that lets you directly access a terminal without all of tmux's downsides, but that takes more time than I have spare right now.

By the way, please don't use JVM flags like this! The ***vast majority*** of JVM flags people throw around as """optimized""" are snake oil. They do basically nothing while potentially causing issues like this! Also, the people behind OpenJDK know what they're doing! The defaults are best for the vast majority of what you'd throw at it. I'd recommend using what's from https://noflags.sh, Dead simple, with help from a Minecraft server admin and modder from the beta days and by an OpenJDK contributor. All it does is switch to using the better GC depending on the JDK version.
